packages/zoe/src/zoeService/installationStorage.js

```diff
@@ -0,0 +1,52 @@
+// @ts-check
+
+import { assert, details as X } from '@agoric/assert';
+import { Far } from '@agoric/marshal';
+import { E } from '@agoric/eventual-send';
+
+export const makeInstallationStorage = () => {
+  /** @type {WeakSet<Installation>} */
+  const installations = new WeakSet();
+
+  /**
+   * Create an installation by permanently storing the bundle. It will be
+   * evaluated each time it is used to make a new instance of a contract.
+   */
+  /** @type {Install} */
+  const install = bundle => {
+    assert(bundle, X`a bundle must be provided`);
```

Comment:
Arguments of messages from the outside may be maliciously malformed, so I always pay more attention to their validation.

It is weird that the only validation we do on the `bundle` is that it is truthy, rejecting `0` or `''` but accepting other things that are trivially invalid. We do no type tests. Not doing more extensive tests here is fine for now. But in that case, accepting `1` and rejecting `0` seems odd. Assuming your real intention was to detect an absent argument, how about

```suggestion
  const install = bundle => {
    assert(bundle !== undefined, X`a bundle must be provided`);
```

Though omitting the thing to install from an `install` call doesn't strike me as more likely than any other mistake. My mod does make the test less precise though, admitting more incorrect things.

I see the type `SourceBundle` is supposed to be a `Record<string, any>`, so if you do want one minimal test, a more precise one is

```suggestion
  const install = bundle => {
    assert.typeof(bundle, 'object', X`a bundle must be provided`);
```

Note that once installations create a zygote, the bundle will be loaded and evaluated, rejecting any that don't do that successfully.